projects/libsodium/secret_key_auth_fuzzer.cc

```diff
@@ -0,0 +1,22 @@
+extern "C" {
+  #include <sodium.h>
+}
+
+class SodiumState {
+public:
+  unsigned char key[crypto_auth_KEYBYTES];
+  unsigned char mac[crypto_auth_BYTES];
+
+  SodiumState() {
+    sodium_init(); // this can fail with a non-zero return code
+    crypto_auth_keygen(key);
```

Comment:
I think this line might cause a problem, but please correct me if I'm wrong about something.

If I understand it correctly, `crypto_auth_keygen` will fill `key` with random bytes.
So the first time I execute this program key can contain one value but the next time I execute the program , key will contain a different value?

I think this will make crashes unreproducible if a crash depends on the value of `key`.

I think a better approach would be creating `SodiumState` on each call to `LLVMFuzzerTestOneInput` and using part of `data` as a seed for `
randombytes_buf_deterministic` instead of using `crypto_auth_keygen`.

Reply:
Yes, I think you are correct. I was only thinking about determinism across a _single_ run of the fuzzer and ensuring reuse of the key across a continuous run of the fuzzer and not multiple runs of the fuzzing program itself. I'll change this and the other fuzzer to generate the keys/nonces using`randombytes_buf_determinstic`.

Reply:
For these tests, I would suggest defining a [custom RNG](https://download.libsodium.org/doc/advanced/custom_rng.html) that returns a deterministic sequence (maybe using a simple counter).

Reply:
I'm not sure about this suggestion (assuming I understand it).
We shouldn't preserve state between calls to `LLVMFuzzerTestOneInput`, since this would lead to calls with the same arguments behaving differently.

Reply:
This sounds like it will work. 
I think defining the RNG and calling `randombytes_set_implementation()` should be done by `LLVMFuzzerTestOneInput` that way the RNG can deterministically return arbitrary values chosen by the fuzzer.

Reply:
Excellent - thanks, I'll work on getting this change in :)